### ISSUE TITLE
DDP doesn't synchronize weights 

### ISSUE THREAD

**Issue by darudimimi** [pytorch/xla#8396]
## 🐛 Bug

I am trying to run a basic example for XLA DDP on a CUDA backend. I have one node with two CUDA devices (RTX 4090). My code (presented below) is similar to https://pytorch.org/xla/master/multi_process_distributed.html. However, it seems that DDP does not synchronize the model weights across threads. To the best of my knowledge during DataParallel training, the model weights should be synchronized after each gradient step.

## To Reproduce
I used the prebuilt docker image according to https://pytorch.org/xla/master/gpu.html#docker.

My .bashrc in Docker:

```
export PATH=$PATH:/usr/local/cuda-12.1/bin
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda-12.1/lib64
export GPU_NUM_DEVICES=2 PJRT_DEVICE=CUDA
```

Here is my code:

```
import os
import sys
import tempfile
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim

from torch.nn.parallel import DistributedDataParallel as DDP

import torch_xla
import torch_xla.core.xla_model as xm
import torch_xla.runtime as xr
import torch_xla.distributed.xla_backend
import torch_xla.distributed.xla_multiprocessing as xmp

import time
import copy

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12357'

    # initialize the xla process group
    dist.init_process_group("xla", rank=rank, world_size=world_size)
    # dist.init_process_group('xla', init_method='xla://')

def cleanup():
    dist.destroy_process_group()

class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net1 = nn.Linear(10, 10)
        self.relu = nn.ReLU()
        self.net2 = nn.Linear(10, 5)

    def forward(self, x):
        return self.net2(self.relu(self.net1(x)))

def demo_basic(rank):
    # xla specific APIs to get rank, world_size.
    new_rank = xr.global_ordinal()
    assert new_rank == rank
    world_size = xr.world_size()
    setup(rank, world_size)

    # create model and move it to XLA device
    device = xm.xla_device()
    print(f"Running basic DDP example on rank {rank}. world_size: {world_size} device: {device}")
    model = ToyModel().to(device)
    ddp_model = DDP(model, gradient_as_bucket_view=True).to(device)

    loss_fn = nn.MSELoss()
    # print([name for name in ddp_model.parameters()])
    optimizer = optim.SGD(ddp_model.parameters(), lr=10)

    for i in range(2):
        optimizer.zero_grad()
        outputs = ddp_model(torch.randn(20, 10).to(device))
        labels = torch.randn(20, 5).to(device)
        loss_fn(outputs, labels).backward()
        print(f'> 1 step: {i} rank: {rank} sum:{model.net1.weight.sum()}')
        optimizer.step()
        xm.mark_step()
        print(f'2 step: {i} rank: {rank} sum:{model.net1.weight.sum()}')

    cleanup()

def run_demo(demo_fn):
    torch_xla.launch(demo_fn, args=())

if __name__ == "__main__":
    run_demo(demo_basic)
```

Output:
```
> python3 ddp_xla.py
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1732025666.903873   28184 coordination_service_agent.cc:340] Coordination agent has successfully connected.
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1732025666.938974   28183 coordination_service_agent.cc:340] Coordination agent has successfully connected.
I0000 00:00:1732025667.044420   28183 service.cc:148] XLA service 0x5ab52f0180f0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1732025667.044471   28183 service.cc:156]   StreamExecutor device (0): NVIDIA GeForce RTX 4090, Compute Capability 8.9
I0000 00:00:1732025667.044972   28183 se_gpu_pjrt_client.cc:943] Using BFC allocator.
I0000 00:00:1732025667.045041   28183 gpu_helpers.cc:114] XLA backend allocating 19043106816 bytes on device 1 for BFCAllocator.
I0000 00:00:1732025667.045081   28183 gpu_helpers.cc:154] XLA backend will use up to 6347702272 bytes on device 1 for CollectiveBFCAllocator.
I0000 00:00:1732025667.056581   28184 service.cc:148] XLA service 0x64d38a1b9820 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1732025667.056645   28184 service.cc:156]   StreamExecutor device (0): NVIDIA GeForce RTX 4090, Compute Capability 8.9
I0000 00:00:1732025667.057513   28184 se_gpu_pjrt_client.cc:943] Using BFC allocator.
I0000 00:00:1732025667.057594   28184 gpu_helpers.cc:114] XLA backend allocating 19045269504 bytes on device 0 for BFCAllocator.
I0000 00:00:1732025667.057664   28184 gpu_helpers.cc:154] XLA backend will use up to 6348423168 bytes on device 0 for CollectiveBFCAllocator.
Running basic DDP example on rank 1. world_size: 2 device: xla:0
Running basic DDP example on rank 0. world_size: 2 device: xla:0
I0000 00:00:1732025667.332156   28507 cuda_dnn.cc:529] Loaded cuDNN version 8902
I0000 00:00:1732025667.343592   28495 cuda_dnn.cc:529] Loaded cuDNN version 8902
> 1 step: 0 rank: 0 sum:-0.6132545471191406
> 1 step: 0 rank: 1 sum:-0.6132545471191406
> 2 step: 0 rank: 1 sum:1.8314828872680664
> 2 step: 0 rank: 0 sum:-0.4641420245170593
[rank1]:[W1119 14:14:28.889105546 logger.cpp:337] Warning: Time stats are currently only collected for CPU and CUDA devices. Please refer to CpuTimer or CudaTimer for how to register timer for other device type. (function operator())
[rank0]:[W1119 14:14:28.889264777 logger.cpp:337] Warning: Time stats are currently only collected for CPU and CUDA devices. Please refer to CpuTimer or CudaTimer for how to register timer for other device type. (function operator())
> 1 step: 1 rank: 0 sum:-0.4641420245170593
> 1 step: 1 rank: 1 sum:1.8314828872680664
> 2 step: 1 rank: 1 sum:-7.215952396392822
> 2 step: 1 rank: 0 sum:2.7132833003997803
I0000 00:00:1732025669.635523   28184 coordination_service_agent.cc:602] Coordination agent has initiated Shutdown().
I0000 00:00:1732025669.639331   28183 coordination_service_agent.cc:602] Coordination agent has initiated Shutdown().
I0000 00:00:1732025669.639746   28184 coordination_service_agent.cc:621] Coordination agent has successfully shut down.
I0000 00:00:1732025669.639779   28183 coordination_service_agent.cc:621] Coordination agent has successfully shut down.

```

1. Run docker container https://pytorch.org/xla/master/gpu.html#docker on a node with multiple CUDA devices
2. Run my code

## Expected behavior

I expected to see the same tensors after a gradient step.
e.g.
```
> 2 step: 1 rank: 1 sum:-7.215952396392822
> 2 step: 1 rank: 0 sum:-7.215952396392822
```

## Environment

 - Reproducible on XLA backend [CPU/TPU/CUDA]:
 - torch_xla version:

**Comment by yaochengji** [comment]
Hi @darudimimi , 

Thanks for submitting the issue. I just submitted a [PR](https://github.com/pytorch/xla/pull/8489) to fix this.
